daceml/onnx/op_implementations/fpga_implementations.py

```diff
@@ -886,6 +886,1034 @@ def make_compute(sdfg, state, vec_width=1):
+                end="")
+            tile_size_m = np.lcm(tile_size_m, vec_width)
+            print(f"{tile_size_m}")
+
+        # fix tile size
+        T = tile_size_m
+
+        if T % output_size_x != 0:
+            raise NotImplementedError(
+                f"Currently tile size ({tile_size_m}) must \
+                be a multiple of the output image width {output_size_x}")
+
+        # Safe delay: this will be used in the compute state, pipeline scope, to insert
+        # a delay between accumulation on the same result if needed.
+        # Further explanations are provided in the compute state.
+
+        # Note: this is a platform and type dependent parameter.
+        L = max(16 - T / vec_width, 0)
+
+        # This implementation uses a flattened nested loop, that overlaps feeding,
+        # computing and draining phases. Each PE is responsible for computing one
+        # tile of one row of the final result C. With the current implementation,
+        # A PE needs K*T cycles to compute the results and then P*T clock cycles
+        # to fully drain them (draining is distributed across PEs).
+        # Therefore, in order to guarantee correctness and deadlock free we have
+        # to ensure that the number of cycles needed to drain the results is less
+        # or equal to the number of cycles needed to compute them.
+        # That is PT <= KT.
+        if K is not None and P > K:
+            raise ValueError(
+                f"Conv Im2Col (tiled): Number of processing elements {P} must be smaller than the K-dimension {K}."
+            )
+
+        def make_read_A(state):
+            '''
+            A is the weights/features streamed to PEs for Im2Col layout i.e. *W*
+            '''
+
+            # A given row of A must be repeated according to B number of tiles
+            # Both N and M can be not a multiple of P and T respectively
+            entry, exit = state.add_map(
+                "read_A",
+                {
+                    "b": f"0:{batch_size}",  # additional batch map
+                    "n0": f"0:ceiling({N}/{P})",
+                    "tm": f"0:ceiling({M}/{T})",
+                    "k": f"0:{K}",
+                    "n1": f"0:{P}"
+                },
+                schedule=dace.ScheduleType.FPGA_Device)
+
+            # The reader of A reads one element per clock cycle.
+            # Note that if P > T+L, then this will be the bottleneck
+            mem = state.add_read("W")
+            pipe = state.add_write("A_pipe")
+
+            # Read data from memory: if we are out-of-bound do not read from memory
+            # but inject dummy data
+            tasklet = state.add_tasklet(
+                "read_A", {"from_memory"}, {"to_kernel"}, f"""\
+data = from_memory if n0 * {P} + n1 < {N} else 0
+to_kernel = data""")
+
+            # Access mapping for Im2Col
+            filter = f"n0 * {P} + n1"  # directly corresponds to rows of Im2Col matrix
```

Comment:
For future use: can you add some explanation about the indexing?
I imagine that in one month,  everyone that should touch this code will spend hours figuring out what's going on here

Reply:
I added a lot more explanations add comments now; in particular:
- Description of general approach in class header
- Description in header of `make_read_W`
- Description of buffering approach in `make_read_X`
- More detailed comments in the index translation sections